strawberryfields/decompositions.py

```diff
@@ -400,6 +400,114 @@ def mach_zehnder(m, n, internal_phase, external_phase, nmax):
     return BS @ Rinternal @ BS @ Rexternal
 
 
+def mach_zehnder_inv(m, n, phi_int, phi_ext, nmax):
+    r"""The inverse of the mach_zehnder matrix."""
+    return np.transpose(np.conj(mach_zehnder(m, n, phi_int, phi_ext, nmax)))
+
+
+def nullMZi(m, n, U):
+    r"""Nullifies element m,n of U using mach_zehnder_inv"""
+    (nmax, mmax) = U.shape
+
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
+
+    if U[m, n] == 0:
+        phi_i = np.pi
+        phi_e = 0
+    elif U[m, n + 1] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = -U[m, n + 1] / U[m, n]
+        phi_i = 2 * np.arctan(np.abs(r))
+        phi_e = -np.angle(r)
+
+    return [n, n + 1, phi_i, phi_e, nmax]
+
+
+def nullMZ(n, m, U):
+    r"""Nullifies element n,m of U using mach_zehnder"""
+    (nmax, mmax) = U.shape
+
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
+
+    if U[n, m] == 0:
+        phi_i = np.pi
+        phi_e = 0
+    elif U[n - 1, m] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = U[n - 1, m] / U[n, m]
+        phi_i = 2 * np.arctan(np.abs(r))
+        phi_e = -np.angle(r)
+
+    return [n - 1, n, phi_i, phi_e, nmax]
+
+
+def rectangular_MZ(V, tol=1e-11):
+    r"""Rectangular decomposition of a unitary matrix, with local
+    phase shifts applied between two interferometers.
+
+    See :ref:`rectangular` or :cite:`clements2016` for more details.
+
+    This function returns a circuit corresponding to an intermediate step in
+    the decomposition as described in Eq. 4 of the article. In this form,
+    the circuit comprises some MZ matrices (as in Eq. 1), then phases on all modes,
+    and more MZ matrices.
+
+    The procedure to construct these matrices is detailed in the supplementary
+    material of the article.
+
+    Args:
+        V (array[complex]): unitary matrix of size n_size
+        tol (float): the tolerance used when checking if the matrix is unitary:
+            :math:`|VV^\dagger-I| \leq` tol
+
+    Returns:
+        tuple[array]: tuple of the form ``(tilist,np.diag(localV),tlist)``
+            where:
+
+            * ``tilist``: list containing ``[n,m,phi_int,phi_ext,n_size]`` of the MZi unitaries needed
+            * ``tlist``: list containing ``[n,m,phi_int,phi_ext,n_size]`` of the MZ unitaries needed
+            * ``localV``: Diagonal unitary sitting sandwiched by MZi's and the MZ's
+    """
+    localV = V
+    (nsize, _) = localV.shape
+
+    diffn = np.linalg.norm(V @ V.conj().T - np.identity(nsize))
```

Comment:
Wonder if it would be good to compare closeness to the identity using `np.allclose`

Reply:
I think the test here is a bit more strict than using `np.allclose`. 
For example `V = U + tol * np.identity(n)`fails the above test, but passes the `np.allclose`

Reply:
That said, good to move this and other decompositions to `np.allclose` as that is easier to read and might be faster? Do you (or @josh146) have any idea about any potential downsides?

Reply:
Good to use `allclose` if possible to keep things consistent (all the tests use the `np.allclose(result, expected, atol=tol, rtol=0)` pattern